src/lib.rs

```diff
@@ -0,0 +1,411 @@
+use near_sdk::json_types::U128;
+use near_sdk::{env, near, require, AccountId, EpochHeight, Timestamp};
+use std::collections::HashMap;
+
+type Balance = u128;
+
+/// Voting contract for any specific proposal. Once the majority of the stake holders agree to
+/// the proposal, the time will be recorded and the voting ends.
+#[near(contract_state)]
+#[derive(Default)]
+pub struct Contract {
+    proposal: String,
+    deadline_timestamp_ms: Timestamp,
+    votes: HashMap<AccountId, Balance>,
+    total_voted_stake: Balance,
+    result: Option<Timestamp>,
+    last_epoch_height: EpochHeight,
+}
+
+// Implement the contract structure
+#[near]
+impl Contract {
+    #[init]
+    #[private]
+    pub fn new(proposal: String, deadline_timestamp_ms: Timestamp) -> Self {
+        require!(!proposal.is_empty(), "Proposal cannot be empty");
+        require!(
+            deadline_timestamp_ms > env::block_timestamp_ms(),
+            "Deadline must be in the future"
+        );
+        Self {
+            proposal,
+            deadline_timestamp_ms,
+            votes: HashMap::new(),
+            total_voted_stake: 0,
+            result: None,
+            last_epoch_height: 0,
+        }
+    }
+
+    /// Method for validators to vote or withdraw the vote.
+    /// Votes for if `is_vote` is true, or withdraws the vote if `is_vote` is false.
+    pub fn vote(&mut self, is_vote: bool) {
+        require!(
```

Comment:
can we remove this assertion since it's already in ping()

Reply:
Nice catch. Updated.